### DISCUSSION TITLE
Use Timers and Scripting together

### DISCUSSION THREAD

**Discussion by dad401** [arendst/Tasmota#13558] [Q&A] [answered]
Is there a way to use Tasmotas 16 timers together with enabled scripting?

I want to use these timers to execute RFSend commands at sunset/sunrise. I know this can be done in a more or less complex way by using a script. But is it possible to use these timers?

Although I compiled my firmware with scripting enabled I can select an action "Rule" under Configuration=>Configure Timer.
But as you know I can only use either Scripting or Rules, so this selection should not be displayed.

**Comment by gemu2015** [comment, marked as answer]
you can set timers with tasmota commands  with the => command from script (see tasmota docs for timer syntax)
then you can read the timer events in the >E section

but it is very easy to add sunrise, sunset to scripts
here is an example of a complex timer script just to see how it can be done

#define USE_TIMERS
#define USE_SCRIPT_WEB_DISPLAY

[Archiv.zip](https://github.com/arendst/Tasmota/files/7470391/Archiv.zip)

**Comment by dad401** [comment]
Thank you for the script. I already found it [on this site](https://forum.creationx.de/forum/index.php?thread/3004-timer-regeln-und-stromausfall/&pageNo=2).

So this concludes that, if I really want to use the 16 timers in the configuration menu, I have to use rules. I can`t use them in any way with scripting?!

Anyway, I adapted your script for my needs (RFSend instead of switching). 
[sockets-timer.txt.gz](https://github.com/arendst/Tasmota/files/7492280/sockets-timer.txt.gz)

**Reply by gemu2015** [reply]
you can use tasmota timers in scripts. you have to catch the events created by the timers in the scripts >E section
 you can choose rules in timer setup of tasmota.
then the >E section of scripter is called with the timer firing
e.g. JSON is send {"Clock":{"Timer":1}}
\D
tvar=0
\>E
tvar=Clock#Timer
reads the timer event into tvar.
if tvar>0 the appropriate timer has fired

**Reply by dad401** [reply]
That's an answer I was looking for - but somehow it's not working.

This is the output with timer 1 enabled:
```
12:33:33.148 RSL: RESULT = {"Timers":"ON","Timer1":{"Enable":1,"Mode":0,"Time":"12:34","Window":0,"Days":"1111111","Repeat":1,"Output":1,"Action":1},"Timer2":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Output":1,"Action":0},"Timer3":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Output":1,"Action":0},"Timer4":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Output":1,"Action":0},"Timer5":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Output":1,"Action":0},"Timer6":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Output":1,"Action":0},"Timer7":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Output":1,"Action":0},"Timer8":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Output":1,"Action":0},"Timer9":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Output":1,"Action":0},"Timer10":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Output":1,"Action":0},"Timer11":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Output":1,"Action":0},"Timer12":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Output":1,"Action":0},"Timer13":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Output":1,"Action":0},"Timer14":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Output":1,"Action":0},"Timer15":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Output":1,"Action":0},"Timer16":{"Enable":0,"Mode":0,"Time":"00:00","Window":0,"Days":"0000000","Repeat":0,"Output":1,"Action":0}}
12:34:00.169 RSL: RESULT = {"POWER":"ON"}
12:34:00.171 RSL: POWER = ON
```
The script gives an error for `tvar=Clock#Timer`: `var not found: Clock#Timer`

The documentation states `If the Tasmota Rules feature has been activated by compiling the code (activated by default in all pre-compiled Tasmota binaries), a rule with Clock#Timer=<timer> will be triggered if written and turned on by the user.` 
So I assume - no Rules feature / Scripting enabled - no Clock#Timer?!

**Reply by gemu2015** [reply]
you don't need rules for that rule option. it also works with scripts.
i guess something is wrong with your script. you should not get var not found.
post your script here so that i can check what is wrong

**Comment by dad401** [comment]
Tried it again with this test script:
```
>D
tvar=0
>E
tvar=Clock#Timer
>S
if tvar>0
then
print tvar is %tvar%
endif
```
Now it works. I had to make sure that action of Timer1 is "Rule" and `tvar=Clock#Timer` is under section `>E`. 

After timer1 has fired, `tvar` stays at 1.00. How do I catch the next Timer1 event (it is a repeatable timer)?
~~Is it just as easy as to reset tvar to 0?~~ It is!
What happens if I have 2 timers which fire at the same time?

**Reply by gemu2015** [reply]
You will get both triggers with different timer numbers

**Comment by dad401** [comment]
This is my script so far. 
```
>D 40
c=0
res=0
tvar=0
s1=0
s2=0
s3=0
s4=0
p:enable=0

>B
res=is1(0 "Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Okt|Nov|Dez|")
res=is(0 "So|Mo|Di|Mi|Do|Fr|Sa|")

#clocksub
if wm==0
then
wcs {s}<hr>{e}
wcs {s}<center><h1 style="color:green;font-size:40px;">%2.0hours%:%2.0mins%:%2.0secs%</h1>{m}%is[wday]% %0day%-%is1[month]% <br>%0year%{e}
wcs {s}<hr>{e}
endif
next
#rf(c)
->RFsend %c%,24,1,15,177

>E
print section E started
tvar=Clock#Timer
if tvar>0 
and enable>0
then
switch tvar
case 1
print timer 1 fired
s1=1
s2=1
s3=1
s4=1

case 2
print timer 2 fired
s1=0
s2=0
s3=0
s4=0

ends
endif
tvar=0

>S
if chg[s1]>0
then
if s1>0
then
=#rf(5575987)
else
=#rf(5575996)
endif
endif
if chg[s2]>0
then
if s2>0
then
=#rf(5576131)
else
=#rf(5576140)
endif
endif
if chg[s3]>0
then
if s3>0
then
=#rf(5576451)
else
=#rf(5576460)
endif
endif
if chg[s4]>0
then
if s4>0
then
=#rf(5584131)
else
=#rf(5584140)
endif
endif

>W
bu(s1 "316-1 on" "316-1 off")bu(s2 "316-2 on" "316-2 off")
bu(s3 "316-3 on" "316-3 off")bu(s4 "316-5 on" "316-5 off")
@ck(enable "Timer rules in script enabled ")
@<hr>
%=#clocksub
```
It's working, but everytime I press a button I defined, section `>E` shows this error `var not found: Clock#Timer`.

**Reply by gemu2015** [reply]
yes that is true. since >E is called on other occasions too when there is no JSON input Clock#Timer.
you simply can ignore that error since tvar is not altered in that case.
